internal/query/roles_v2.go

```diff
@@ -0,0 +1,668 @@
+package query
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+
+	pb "github.com/authzed/authzed-go/proto/authzed/api/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+
+	"go.infratographer.com/permissions-api/internal/iapl"
+	"go.infratographer.com/permissions-api/internal/storage"
+	"go.infratographer.com/permissions-api/internal/types"
+)
+
+// V2 Role and Role Bindings
+
+func (e *engine) namespaced(name string) string {
+	return e.namespace + "/" + name
+}
+
+func (e *engine) CreateRoleV2(ctx context.Context, actor, owner types.Resource, roleName string, actions []string) (types.Role, error) {
+	ctx, span := e.tracer.Start(ctx, "engine.CreateRoleV2")
+
+	defer span.End()
+
+	roleName = strings.TrimSpace(roleName)
+
+	role := newRoleWithPrefix(e.schemaTypeMap[e.rbac.RoleResource.Name].IDPrefix, roleName, actions)
+	roleRels := e.roleV2Relationships(role)
+	roleRels = append(roleRels, e.roleV2OwnerRelationship(role, owner)...)
+
+	dbCtx, err := e.store.BeginContext(ctx)
+	if err != nil {
+		return types.Role{}, nil
+	}
+
+	dbRole, err := e.store.CreateRole(dbCtx, actor.ID, role.ID, roleName, owner.ID)
+	if err != nil {
+		return types.Role{}, err
+	}
+
+	request := &pb.WriteRelationshipsRequest{Updates: roleRels}
+
+	if _, err := e.client.WriteRelationships(ctx, request); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		return types.Role{}, err
+	}
+
+	if err = e.store.CommitContext(dbCtx); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		logRollbackErr(e.logger, e.store.RollbackContext(dbCtx))
+
+		// No rollback of spicedb relations are done here.
+		// This does result in dangling unused entries in spicedb,
+		// however there are no assignments to these newly created
+		// and now discarded roles and so they won't be used.
```

Comment:
there will be dnagling relationships, nothing other than that would happen since `GetRoleV2` will fail if the role doesn't exist in the permissions-api DB, so as far as the users' concern, no such role exists

Reply:
role-binding would failed as (my current implementation) also checks the permission-api DB for the role